purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt

```diff
@@ -0,0 +1,211 @@
+          "customer_info_request_date": 1684245988766
+        }
+    """.trimIndent()
+    private val cacheKeyPrefix = "com.revenuecat.purchases.${Constants.apiKey}"
+    private val productEntitlementMappingCacheKey = "$cacheKeyPrefix.productEntitlementMapping"
+    private val initialActiveTransaction = StoreTransactionFactory.createStoreTransaction(
+        skus = listOf(testProductId),
+        purchaseToken = "test-token"
+    )
+    private val initialActivePurchases = mapOf(
+        initialActiveTransaction.purchaseToken.sha1() to initialActiveTransaction
+    )
+
+    @Before
+    fun setup() {
+        setupMockServer()
+    }
+
+    @After
+    fun tearDown() {
+        tearDownTest()
+        server.shutdown()
+        Purchases.proxyURL = null
+    }
+
+    @Test
+    fun entersOfflineEntitlementsModeIfNoCachedCustomerInfoAndCustomerInfoRequestReturns500() {
+        ensureBlockFinishes { latch ->
+            setupTest(
+                buildSharedPreferencesMap(shouldIncludeCustomerInfo = false),
+                initialActivePurchases
+            ) {
+                Purchases.sharedInstance.getCustomerInfoWith(
+                    fetchPolicy = CacheFetchPolicy.FETCH_CURRENT,
+                    onError = {
+                        latch.countDown()
+                        fail("Expected success but got error: $it")
+                    },
+                    onSuccess = { receivedCustomerInfo ->
+                        assertThat(receivedCustomerInfo.entitlements.active.keys).containsExactly(
+                            "pro_cat"
+                        )
+                        latch.countDown()
+                    }
+                )
+            }
+        }
+    }
+
+    @Test
+    fun doesNotEnterOfflineEntitlementsModeIfCachedCustomerInfoAndCustomerInfoRequestReturns500() {
+        ensureBlockFinishes { latch ->
+            setupTest(
+                buildSharedPreferencesMap(shouldIncludeCustomerInfo = true),
+                initialActivePurchases
+            ) {
+                Purchases.sharedInstance.getCustomerInfoWith(
+                    fetchPolicy = CacheFetchPolicy.FETCH_CURRENT,
+                    onError = {
+                        latch.countDown()
+                    },
+                    onSuccess = {
+                        latch.countDown()
+                        fail("Expected error but got success: $it")
+                    }
+                )
+            }
+        }
+    }
+
+    @Test
+    fun entersOfflineEntitlementsModeIfPurchaseRequestReturns500() {
+        val storeProduct = StoreProductFactory.createGoogleStoreProduct(
+            productId = testProductId,
+            basePlanId = testBasePlanId,
+            subscriptionOptionsList = listOf(
+                StoreProductFactory.createGoogleSubscriptionOption(
+                    productId = testProductId,
+                    basePlanId = testBasePlanId,
+                )
+            )
+        )
```

Comment:
Creating these fake products is a bit hacky... We could create them from mocked google product details but that also doesn't feel great to me though, I think this is relatively simpler.